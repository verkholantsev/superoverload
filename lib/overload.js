'use strict';


var getType = require('./getType');
var pair = require('./pair');
module.exports = overload;

/**
 * Function overload implementation
 *
 * Takes signatures and functions as arguments, like this:
 *
 * var fn = overload(
 *     ['number'],
 *     function (a) { return 'It is a number'; },
 *
 *     ['string'],
 *     function (a) { return 'It is a string'; }
 * )
 *
 * fn(1); // => 'It is a number'
 * fn(''); // => 'It is a string'
 *
 * Returns an overloaded function, that will call the function with corresponding signature
 *
 * Fallback function can be passed as a first argument, like this:
 *
 * var fn = overload(
 *     function (a) { return 'It is something else'; },
 *
 *     ['number'],
 *     function (a) { return 'It is a number'; }
 * )
 *
 * fn(1); // => 'It is a number'
 * fn(''); // => 'It is something else'
 *
 * @return {function}
 */
function overload() {
    var args = new Array(arguments.length);
    for(var i = 0; i < args.length; ++i) {
        args[i] = arguments[i];
    }
    if (args.length % 2 > 0) {
        var defaultFn = args.shift();
    }
    var pairs = pair(args);
    return function () {
        var args = new Array(arguments.length);
        for(var i = 0; i < args.length; ++i) {
            args[i] = arguments[i];
        }
        var pair = pairs.filter(function (pair) {
            var signature = pair[0];
            return args.length === signature.length && args.every(function (arg, index) {
                return getType(arg) === signature[index];
            });
        })[0];
        if (!pair) {
            if (!defaultFn) {
                var description = args.map(getType).join(', ');
                throw new Error('No matching function for call with signature "' + description + '"');
            }

            return defaultFn.apply(this, arguments);
        }
        var fn = pair[1];
        return fn.apply(this, arguments);
    };
}
