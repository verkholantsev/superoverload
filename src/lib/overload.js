'use strict';

import getType from './getType';
import pair from './pair';

/**
 *
 * @param {array} array
 * @return {string}
 */
function serializeSignature(array) {
    return array
        .map(function(arg, index) {
            return '_' + index;
        })
        .join(',');
}

/**
 * Function overload implementation
 *
 * Takes signatures and functions as arguments, like this:
 *
 * const fn = overload(
 *     ['number'],
 *     function (a) { return 'It is a number'; },
 *
 *     ['string'],
 *     function (a) { return 'It is a string'; }
 * )
 *
 * fn(1); // => 'It is a number'
 * fn(''); // => 'It is a string'
 *
 * Returns an overloaded function, that will call the function with corresponding signature
 *
 * Fallback function can be passed as a first argument, like this:
 *
 * const fn = overload(
 *     function (a) { return 'It is something else'; },
 *
 *     ['number'],
 *     function (a) { return 'It is a number'; }
 * )
 *
 * fn(1); // => 'It is a number'
 * fn(''); // => 'It is something else'
 *
 * @return {function}
 */
export default function overload(...args) {
    const defaultFn = args.length % 2 > 0 ? args.shift() : null;
    const pairs = pair(args);
    const fns = new Array(pairs.length);
    let longestSignature = [];

    const ifs = pairs
        .map((pair, index) => {
            const sample = 'if (hashKey === "%signature%") { return fns[%index%].call(this, %args%); }';
            const signature = pair[0];
            const fn = pair[1];
            const hashKey = signature.join(', ');

            fns[index] = fn;
            if (signature.length > longestSignature.length) {
                longestSignature = signature;
            }

            return sample
                .replace('%signature%', hashKey)
                .replace('%index%', index)
                .replace('%args%', serializeSignature(signature));
        })
        .join(' else ');

    const serializedSignature = serializeSignature(longestSignature);
    const code = [
            'return function(' + serializedSignature + ') {',
                'var hashKey = "";',
                'for (var i = 0, len = arguments.length; i < len; i++) {',
                    'hashKey += getType(arguments[i]);',
                    'if (i !== len - 1) {',
                        'hashKey += ", ";',
                    '}',
                '}',
                ifs,

                pairs.length > 0 ? ' else { ' : '',

                'if (!defaultFn) {',
                    'throw new Error(\'No matching function for call with signature "\' + hashKey + \'"\');',
                '}',

                pairs.length > 0 ? '}' : '',
                'return defaultFn.apply(this, arguments);',
            '}'
        ];

    /* jshint evil: true */
    const superFunc = new Function('getType, fns, defaultFn', code.join(''));
    /* jshint evil: false */

    return superFunc(getType, fns, defaultFn);
}
