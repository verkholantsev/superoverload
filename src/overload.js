// @flow

import getType from './get-type';
import pair from './pair';
import serializeSignature from './serialize-signature';
import getUnsupportedTypes from './get-unsupported-types';

/**
 * Function overload implementation
 *
 * Takes signatures and functions as arguments, like this:
 *
 * const fn = overload(
 *     ['number'],
 *     a =>'It is a number',
 *
 *     ['string'],
 *     a => 'It is a string'
 * )
 *
 * fn(1); // => 'It is a number'
 * fn(''); // => 'It is a string'
 *
 * Returns an overloaded function, that will call the function with corresponding signature
 *
 * Fallback function can be passed as a first argument, like this:
 *
 * const fn = overload(
 *     a => 'It is something else',
 *
 *     ['number'],
 *     a => 'It is a number'
 * )
 *
 * fn(1); // => 'It is a number'
 * fn(''); // => 'It is something else'
 */
export default function overload(...args: Array<*>): Function {
    const defaultFn = args.length % 2 > 0 ? args.shift() : null;
    const pairs = pair(args);

    const ifsArray = new Array(pairs.length);
    const fns = new Array(pairs.length);
    let longestSignature = [];

    for (let i = 0; i < pairs.length; i++) {
        const pair = pairs[i];
        const signature = pair[0];
        const fn = pair[1];

        const unsupportedTypes = getUnsupportedTypes(signature);
        if (unsupportedTypes.length > 0) {
            throw new Error(
                `Signature "${signature.join(', ')}" contains unsupported types: "${unsupportedTypes.join(', ')}"`
            );
        }

        const hashKey = signature.join(', ');
        fns[i] = fn;

        if (signature.length > longestSignature.length) {
            longestSignature = signature;
        }

        ifsArray[i] = `
if (hashKey === '${hashKey}') {
    return fns[${String(i)}].call(this, ${serializeSignature(signature)});
}`;
    }

    const ifs = ifsArray.join(' else ');

    const serializedSignature = serializeSignature(longestSignature);
    const code = `
return function overloadedFn(${serializedSignature}) {
    var hashKey = '';
    var len = arguments.length;
    var args = new Array(len);

    for (var i = 0; i < len; i++) {
        args[i] = arguments[i];
    }

    for (var i = 0; i < len; i++) {
        hashKey += getType(args[i]);
        if (i !== len - 1) {
            hashKey += ', ';
        }
    }
    ${ifs}
    ${pairs.length > 0 ? 'else {' : ''}
    if (!defaultFn) {
        throw new Error('No matching function for call with signature "' + hashKey + '"');
    }
    ${pairs.length > 0 ? '}' : ''}
    return defaultFn.apply(this, arguments);
}`;

    const superFunc = new Function('getType, fns, defaultFn', code);

    // $FlowFixMe errors with `new Function(...)`
    return superFunc(getType, fns, defaultFn);
}
