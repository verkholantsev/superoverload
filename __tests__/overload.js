'use strict';

import overload from '../src/overload';
import prettier from 'prettier';

const expectFormattedFnToMatchSnapshot = async fn => {
    const config = Object.assign({ parser: 'babylon' }, await prettier.resolveConfig(__filename));
    expect(prettier.format(fn.toString(), config)).toMatchSnapshot();
};

describe('overload()', () => {
    it('should return a function', () => {
        expect(overload()).toEqual(expect.any(Function));
    });

    describe('called with even number of arguments', () => {
        describe('should return function, which', () => {
            let fn;
            beforeEach(() => {
                fn = overload(
                    ['string'],
                    () => 'string',

                    ['array'],
                    () => 'array'
                );
            });

            it('calls function with corresponding signature', () => {
                expect(fn('string')).toEqual('string');
                expect(fn([])).toEqual('array');
            });

            it('result of toString() matches snapshot', async () => {
                await expectFormattedFnToMatchSnapshot(fn);
            });

            describe('throw an exception', () => {
                it('if there is no function with corresponding signature', () => {
                    const message = 'No matching function for call with signature "number, number"';
                    expect(() => fn(1, 2)).toThrow(message);
                });
            });
        });
    });

    describe('called with odd number of arguments', () => {
        describe('should return function, which', () => {
            let fn;
            beforeEach(() => {
                fn = overload(
                    () => 'default',

                    ['string'],
                    () => 'string',

                    ['number'],
                    () => 'number'
                );
            });

            it('calls function with corresponding signature', () => {
                expect(fn('string')).toEqual('string');
                expect(fn(1)).toEqual('number');
            });

            it('result of toString() matches snapshot', async () => {
                await expectFormattedFnToMatchSnapshot(fn);
            });

            describe('calls first function', () => {
                it('if there is no function with corresponding signature', () => {
                    expect(fn(true)).toEqual('default');
                });

                describe('with', () => {
                    beforeEach(() => {
                        fn = overload(arg => arg);
                    });

                    it('all passed arguments', () => {
                        expect(fn(true)).toEqual(true);
                    });
                });
            });
        });
    });

    describe('called with complicated signature', () => {
        let fn;
        beforeEach(() => {
            fn = overload(['number', 'string', 'array', 'object', 'function', 'regexp', 'date'], () => 'ok');
        });

        it('should return working overloaded function', () => {
            expect(fn(1, '', [], {}, () => {}, /1/, new Date())).toEqual('ok');
        });
    });

    describe('called with unsupported types', () => {
        it('should throw an error', () => {
            expect(() => overload(['number', 'string', 'unsupported-type'], () => {})).toThrow(
                'Signature "number, string, unsupported-type" contains unsupported types: "unsupported-type"'
            );
        });
    });
});
