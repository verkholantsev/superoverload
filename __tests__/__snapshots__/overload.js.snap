// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`overload() called with even number of arguments should return function, which result of toString() matches snapshot 1`] = `
"function overloadedFn(_0) {
    var hashKey = '';
    var len = arguments.length;
    var args = new Array(len);

    for (var i = 0; i < len; i++) {
        args[i] = arguments[i];
    }

    for (var i = 0; i < len; i++) {
        hashKey += getType(args[i]);
        if (i !== len - 1) {
            hashKey += ', ';
        }
    }

    if (hashKey === 'string') {
        return fns[0].call(this, _0);
    } else if (hashKey === 'array') {
        return fns[1].call(this, _0);
    } else {
        if (!defaultFn) {
            throw new Error('No matching function for call with signature \\"' + hashKey + '\\"');
        }
    }
    return defaultFn.apply(this, arguments);
}
"
`;

exports[`overload() called with odd number of arguments should return function, which result of toString() matches snapshot 1`] = `
"function overloadedFn(_0) {
    var hashKey = '';
    var len = arguments.length;
    var args = new Array(len);

    for (var i = 0; i < len; i++) {
        args[i] = arguments[i];
    }

    for (var i = 0; i < len; i++) {
        hashKey += getType(args[i]);
        if (i !== len - 1) {
            hashKey += ', ';
        }
    }

    if (hashKey === 'string') {
        return fns[0].call(this, _0);
    } else if (hashKey === 'number') {
        return fns[1].call(this, _0);
    } else {
        if (!defaultFn) {
            throw new Error('No matching function for call with signature \\"' + hashKey + '\\"');
        }
    }
    return defaultFn.apply(this, arguments);
}
"
`;
